pkg/core/transaction/cosigner.go

```diff
@@ -0,0 +1,58 @@
+package transaction
+
+import (
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// The maximum number of AllowedContracts or AllowedGroups
+const maxSubitems = 16
+
+// Cosigner implements a Transaction cosigner.
+type Cosigner struct {
+	Account          util.Uint160      `json:"account"`
+	Scopes           WitnessScope      `json:"scopes"`
+	AllowedContracts []util.Uint160    `json:"allowedContracts,omitempty"`
+	AllowedGroups    []*keys.PublicKey `json:"allowedGroups,omitempty"`
+}
+
+// EncodeBinary implements Serializable interface.
+func (c *Cosigner) EncodeBinary(bw *io.BinWriter) {
+	bw.WriteBytes(c.Account[:])
+	bw.WriteB(byte(c.Scopes))
+	if c.Scopes&CustomContracts != 0 {
+		bw.WriteVarUint(uint64(len(c.AllowedContracts)))
+		for _, c := range c.AllowedContracts {
+			bw.WriteBytes(c[:])
+		}
+	}
+	if c.Scopes&CustomGroups != 0 {
+		bw.WriteArray(c.AllowedGroups)
+	}
+}
+
+// DecodeBinary implements Serializable interface.
+func (c *Cosigner) DecodeBinary(br *io.BinReader) {
+	br.ReadBytes(c.Account[:])
+	c.Scopes = WitnessScope(br.ReadB())
+	if c.Scopes&CustomContracts != 0 {
+		contentsLength := br.ReadVarUint()
+		if contentsLength > maxSubitems {
+			br.Err = fmt.Errorf("cosigner can't contain more than %v allowed contracts", maxSubitems)
+		}
+		c.AllowedContracts = make([]util.Uint160, contentsLength)
+		for i := range c.AllowedContracts {
+			br.ReadBytes(c.AllowedContracts[i][:])
+		}
+	} else {
+		c.AllowedContracts = []util.Uint160{}
```

Comment:
We can, but this would differs from c# implementation (see https://github.com/neo-project/neo/blob/master/src/neo/Network/P2P/Payloads/Cosigner.cs#L36). So, do we have to live `nil` here?

Reply:
We need to reproduce some behaviour associated with these values, not specific implementation details. There has to be a reason for such initialization (like when you have to have initialized slice for some code to work), but if we can live with these fields being `nil`, I think we should do so, useless allocations don't make anyone happy.